app/lib/actions/remote_execution/run_proxy_command.rb

```diff
@@ -23,12 +66,62 @@ def rescue_strategy
       end
 
       def failed_run?
-        exit_status && proxy_output[:exit_status] != 0
+        output[:result] == 'initialization_error' ||
+          (exit_status && proxy_output[:exit_status] != 0)
       end
 
       def exit_status
         proxy_output && proxy_output[:exit_status]
       end
+
+      def metadata
+        output[:metadata]
+      end
+
+      def metadata=(thing)
+        output[:metadata] = thing
+      end
+
+      private
+
+      def default_connection_options
+        # Fails if the plan is not finished within 60 seconds from the first task trigger attempt on the smart proxy
+        # If the triggering fails, it retries 3 more times with 15 second delays
+        { :connection_options => { :retry_interval => 15, :retry_count => 4, :timeout => 60 } }
+      end
+
+      def handle_connection_exception(exception, event = nil)
+        output[:metadata] ||= {}
+        metadata[:failed_proxy_tasks] ||= []
+        options = input[:connection_options]
+        metadata[:failed_proxy_tasks] << format_exception(exception)
+        output[:proxy_task_id] = nil
+        if metadata[:failed_proxy_tasks].count < options[:retry_count]
+          suspend do |suspended_action|
+            @world.clock.ping suspended_action,
+                              Time.now + options[:retry_interval],
+                              event
+          end
+        else
+          raise exception
+        end
+      end
+
+      def format_exception(exception)
+        { output[:proxy_task_id] =>
+          { :exception_class => exception.class.name,
+            :execption_message => exception.message } }
+      end
+
+      def with_connection_error_handling(event = nil)
```

Comment:
Shouldn't this be part of foreman-tasks, as it's more about the connection between forman-tasks and smart-proxy-dynflow, than about the remote execution itself?